rabid.8
.TH "rabid" "8" "0.70" "russell@coker.com.au" "Postal"
.SH "NAME"
rabid \- program to test POP server throughput.

.SH "SYNOPSIS"
.B rabid
.I [\-r max\-connections\-per\-minute] [\-p processes] [\-l local\-address]
.I [\-c messages\-per\-connection] [\-a] [\-s ssl\-percentage] [\-i imap\-percentage]
.I [\-b qmail\-pop] [\-d download\-percentage[:delete\-percentage]]
.I [\-[z|Z] debug\-file] [\-u]
.I pop\-server user\-list\-filename

.SH "DESCRIPTION"
This manual page documents briefly the
.BR rabid ,
program.
.P
It is designed to test the performance of POP email servers by reading all
messages from randomly selected accounts as fast as possible.  A future version
will support rate limiting to provide a constant load (for testing SMTP
servers).
.P
The
.B pop\-server
parameter specifies the IP address or name of the mail server that the mail
is to downloaded from.  If you want to specify a port other than port 110 then
enclose the host address in
square brackets and have the port address immidiately following.  If you want
a DNS lookup for every connection (for testing round\-robin DNS) then
immediately preceed the host address with a '+' character.
.P
The
.B user\-list\-filename
is the name of a file which contains a list of user's email addresses and
passwords.  It will have one address per line and the password follows the
address with a space to seperate.
.P
The
.B processes
parameter is the number of processes that should be forked off to attempt
seperate connections.  A well configured mail server won't accept an unlimited
number of connections so make sure you don't specify a number larger than the
number your mail server is configured to handle.  Also for sensible results
make sure that you don't use enough to make your server thrash as the results
won't be representative of real\-world use.
.P
The
.B max\-connections\-per\-minute
parameter is for limiting the number of connections that the program
makes.  This is designed to be used when you want to test the performance
of other programs when the system is under load.  The default is 24000
connections per minute.
.P
The
.B messages\-per\-connection
parameter specifies the maximum number of messages to download in a single POP
session.  The default is \-1 (unlimited).
.P
The
.B local\-address
parameter specifies which local IP address(es) are used to make the outbound
connections.  Specified in the same way as the remote address.  This is good
for testing LocalDirectors or other devices that perform differently depending
on which source IP address was used.
.P
The
.B \-a
command turns on all logging.  All message data received will be logged.  This
will make it slow and it may not be able to saturate a fast Ethernet link...
.P
The
.B \-s
switch specifies the percentage of connections which are to use
.B TLS
AKA
.B SSL.
Use 0 for no SSL, or 100 for always SSL, or any number in between.  Default is
0.
.P
The
.B \-i
switch specifies the percentage of IMAP connections (default is POP).
.P
The
.B \-b
switch allows you to specify breakage strings.  Currently the only option is
for Qmail POP server which adds an extra blank line at the end of each
message.
.B \-b qmail\-pop
means to not report this as an error.
.P
.B \-d download\-percentage[:delete\-percentage]
allows you to specify what percentage of the messages are downloaded and what
percentage of the downloaded messages are deleted.  Default is 100%.
.P
The
.B \-u
switch causes the domain of user-names to be ignored.  This allows you to have
a single file with user-names and passwords which can be used by postal and
rabid when using a server which doesn't accept a domain.  By default postal
will ignore the password field, rabid may or may not need the domain depending
on the configuration of the POP server.  The default is to use the domain
(which is required if the same user is present in multiple domains), this
switch causes the domain part to be stripped from the user-name field.
.P
The
.B \-z
switch allows you to specify a debugging file base.  From this base one file
is created for each thread (with a ':' and the thread number appended), each
file is used to log all IO performed by that thread for debugging purposes.
.P
The
.B \-Z
switch is the same but creates a separate file for each connection as well
with an attitional ':' appended followed by the connection number.

.SH "BUGS"
Doesn't actually do SSL or IMAP yet.

.SH "RETURN CODES"
.TP
.B 0
No Error
.TP
.B 1
Bad Parameters
.TP
.B 2
System Error, lack of memory or some other resource

.SH "AUTHOR"
This program, it's manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>.

.SH "AVAILABILITY"
The source is available from http://doc.coker.com.au/projects/postal/ .
.P
See http://etbe.coker.com.au/category/benchmark for further information.

.SH "SEE ALSO"
.BR postal (8), bhm (8)


debian/changelog
postal (0.72) unstable; urgency=low

  * Make LMTP work.
  * Accept a tab as a field delimiter for the user-list file.

 -- Russell Coker <russell@coker.com.au>  Sat, 14 Jan 2012 19:27:21 +1100

postal (0.71) unstable; urgency=low

  * Made rabid get the password right when -u is used.  Thanks to Lars Nordin
    for the patch.
  * Made rabid correctly check for the '.' at the end of the results of the
    list command.
  * Made all the programs correctly handle a userlist that has empty lines.
  * Use the -lgcrypt option when using gcrypt.
  * Only call gcry_control() if GNUTLS_VERSION_NUMBER <= 0x020b00
  * Ignore blank lines in user list.
  * Add subject, date, and msgid to the md5sum in the client.
    Closes: #611907
  * Stop using gnutls_kx_set_priority()
    Closes: #624038
  * Added a patch from Michael Brown <michael@netdirect.ca> to make an RPM build
    on RHEL5 (and probably other RPM systems).
  * Added a patch from Brad Hards <bradh@frogmouth.net> to fix typos in the
    man page.
  * Rewrote a lot of the md5 checking code, now the md5 hash is stored in a
    header.  It should work now.

 -- Russell Coker <russell@coker.com.au>  Mon, 02 Jan 2012 17:08:50 +1100

postal (0.70) unstable; urgency=low

  * Made it build correctly when no SSL support is configured.  Also tidy
    up the documentation.

 -- Russell Coker <russell@coker.com.au>  Mon, 26 May 2008 19:33:48 +1000

postal (0.69) unstable; urgency=low

  * Really fix the GCC 4.3 issues.  The main change was to remove the
    OS/2 and Win32 support (which probably didn't work anyway).  The previous
    version actually fixed GCC 4.2 issues.  I don't plan to fix the
    ext/hash_map issue yet, it doesn't stop anything from working and I'm
    afraid of breaking things for the majority of systems which use older
    headers.  Thanks to Bradley Smith and Chris Lamb for sending a patch.
    Closes: #474831
    Thanks to Martin Michlmayr for sending a patch.
    Closes: #417480

 -- Russell Coker <russell@coker.com.au>  Sat, 19 Apr 2008 08:07:00 +1000

postal (0.68) unstable; urgency=low

  * Added $(DESTDIR) support to the Makefile and -lrt for nanosleep() on
    OpenSolaris (the OpenSolaris code is not tested).
  * Fixed a minor compilation bug when not using GNUTLS.
    Thanks Frank Fegert <frank.fegert@web.de>.
  * Made it not use C++ warning flags on C compilation.
  * Updated it to the latest Debian standards for Lenny and compile with
    GCC 4.3. Closes: #474831

 -- Russell Coker <russell@coker.com.au>  Wed,  9 Apr 2008 11:16:00 +1000

postal (0.67-0.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with gcc-4.3. Thanks to MArtin Michlmayr. Closes: #417480

 -- Andreas Barth <aba@not.so.argh.org>  Sun, 16 Mar 2008 23:48:44 +0000

postal (0.67) unstable; urgency=low

  * Changed license to GPLv3.

 -- Russell Coker <russell@coker.com.au>  Mon,  2 Jul 2007 20:12:00 +1000

postal (0.66) unstable; urgency=low

  * Added MessageId to Postal.
  * Added "make dep".
  * Made GNUTLS work in bhm.

 -- Russell Coker <russell@coker.com.au>  Fri,  2 Mar 2007 18:44:00 +1100

postal (0.65) unstable; urgency=low

  * Added new program bhm to listen on port 25 and send mail to /dev/null.
    This allows testing mail relay systems.
  * Fixed a minor bug in reporting when compiled without SSL.
  * Made postal write the date header in correct RFC2822 format.
  * Removed the name-expansion feature, it confused many people and is not
    needed now that desktop machines typically have 1G of RAM.
    Now postal and rabid can have the same user-list file.
  * Moved postal-list into the bin directory.
  * Changed the thread stack size to 32K (used to be the default of 10M) to
    save virtual memory size (not that this makes much difference to
    anything other than the maximum number of threads on i386).
  * Added a minimum message size option to Postal.
  * Added a Postal option to specify a list of sender addresses separately to
    the list of recipient addresses.
  * Removed some unnecessary error messages.
  * Handle EINTR to allow ^Z and "bg" from the command line.  I probably
    don't handle all cases, but now that I agree that failure to handle ^Z is
    an error I expect bug reports.
  * Made the test programs display output on the minute, previously they
    displayed once per minute (EG 11:10:35) while now it will be 11:10:00.
    This also means that the first minute reported will have something less
    than 60 seconds of data - this does not matter as a mail server takes
    longer than that to get up to speed.
  * Added support for GNUTLS and made the Debian package build with it.
    Note that BHM doesn't yet work correctly with TLS.
  * Made the programs exit cleanly.

 -- Russell Coker <russell@coker.com.au>  Thu, 18 Jan 2007 08:14:00 +1100

postal (0.62.1) unstable; urgency=low

  * NMU
  * Build-Depend: libssl-dev, not libssl096-dev.  Closes: #176956

 -- LaMont Jones <lamont@debian.org>  Sun, 25 May 2003 21:33:07 -0600

postal (0.62) unstable; urgency=high

  * Added an INSTALL document.

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Tue, 14 Jan 2003 13:28:00 +0100

postal (0.61-1) unstable; urgency=high

  * Changed the section to main.

 -- Russell Coker <russell@coker.com.au>  Sun, 16 Jun 2002 23:17:00 +0200

postal (0.61) unstable; urgency=high

  * Fixed the makefile to not use "-d" or "-p" options to $(INSTALL).

  * Changed connect() functions to Connect() because of some problems with
    Solaris.

  * Removed the optind=0 which caused problems on BSD.

 -- Russell Coker <russell@coker.com.au>  Sun, 16 Jun 2002 21:16:00 +0200

postal (0.60) unstable; urgency=high

  * Added #include <stdlib.h> to mutex.cpp to make exit() be resolved correctly.

  * Fixed the delete percentage code for rabid.  The option "-d100:0" will mean
    that no messages get deleted.

  * Added a new version of the thread code from Bonnie++, fixes a few bugs.


rabid.cpp
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include "userlist.h"
#include "client.h"
#include <cstdlib>
#include <unistd.h>
#include <sys/wait.h>
#include <signal.h>
#include <stdio.h>
#include <strings.h>
#include "postal.h"
#include "logit.h"
#ifdef USE_GNUTLS
#include <errno.h>
#include <gcrypt.h>
GCRY_THREAD_OPTION_PTHREAD_IMPL;
#endif

void usage()
{
  printf("Usage: rabid [-r max-connections-per-minute] [-p processes] [-l local-address]\n"
         "             [-c messages-per-connection] [-a] [-i imap-percentage]\n"
#ifdef USE_SSL
         "             [-s ssl-percentage] [-d download-percentage[:delete-percentage]]\n"
#endif
         "             [-[z|Z] debug-file] [-u]\n"
         "             pop-server user-list-filename\n"
         "\n"
         "Rabid Version: " VER_STR "\n");
  exit(eParam);
}

int exitCount = 0;

void endit(int)
{
  exitCount++;
  if(exitCount > 2)
    exit(1);
}

int main(int argc, char **argv)
{

  int processes = 1;
  int connectionsPerMinute = 0;
  int msgsPerConnection = -1;
  const char *ourAddr = NULL;
  bool logAll = false;
#ifdef USE_SSL
  int ssl = 0;
#endif
  int imap = 0;
  int downloadPercent = 100, deletePercent = 100;
  TRISTATE qmail_pop = eNONE;
  PCCHAR debugName = NULL;
  bool debugMultipleFiles = false;
  bool strip_domain = false;
#ifdef USE_GNUTLS
#if GNUTLS_VERSION_NUMBER <= 0x020b00
  gcry_control(GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
#endif
  gnutls_global_init();
  if(!gnutls_check_version(GNUTLS_VER))
  {
    fprintf(stderr, "Needs version " GNUTLS_VER " of GNUTLS\n");
    exit(1);
  }
#endif

  int c;
  while(-1 != (c = getopt(argc, argv, "ab:d:c:i:l:p:r:s:uz:Z:")) )
  {
    switch(char(c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'a':
        logAll = true;
      break;
      case 'b':
        if(!strcasecmp(optarg, "qmail-pop"))
          qmail_pop = eWONT;
      break;
      case 'c':
        msgsPerConnection = atoi(optarg);
      break;
      case 'd':
        if(sscanf(optarg, "%d:%d", &downloadPercent, &deletePercent) < 2)
          deletePercent = 100;
      break;
      case 'i':
        imap = atoi(optarg);
      break;
      case 'l':
        ourAddr = optarg;
      break;
      case 'p':
        processes = atoi(optarg);
      break;
      case 'r':
        connectionsPerMinute = atoi(optarg);
      break;
      case 's':
#ifdef USE_SSL
        ssl = atoi(optarg);
#else
        usage();
#endif
      break;
      case 'u':
        strip_domain = true;
      break;
      case 'Z':
        debugMultipleFiles = true;
      case 'z':
        debugName = optarg;
      break;
    }
  }
  if(processes < 1 || processes > MAX_PROCESSES || connectionsPerMinute < 0)
    usage();
#ifdef USE_SSL
  if(ssl < 0 || ssl > 100)
    usage();
#endif
  if(imap < 0 || imap > 100)
    usage();
  if(downloadPercent < 0 || downloadPercent > 100)
    usage();
  if(deletePercent < 0 || deletePercent > 100)
    usage();
  if(optind + 2 != argc)
    usage();

  UserList ul(argv[optind + 1], true, strip_domain);

  struct sigaction sa;
  sa.sa_sigaction = NULL;
  sa.sa_handler = SIG_IGN;
  sa.sa_flags = 0;
  if(sigaction(SIGPIPE, &sa, NULL))
  {
    printf("Can't block SIGPIPE.\n");
    return eSystem;
  }

  sa.sa_flags = SA_SIGINFO;
  sa.sa_handler = &endit;
  if(sigaction(SIGINT, &sa, NULL))
  {
    printf("Can't handle SIGINT.\n");
    return eSystem;
  }

  printf("time,messages,data(K),errors,connections"
#ifdef USE_SSL
         ",SSL connections"
#endif
         ",IMAP connections\n");
  Logit log("rabid.log", logAll, false, 0);
  Logit *debug = NULL;
 
  if(debugName)
    debug = new Logit(debugName, false, debugMultipleFiles, 0);
  client popper(&exitCount, argv[optind], ourAddr, ul, processes, msgsPerConnection, &log
#ifdef USE_SSL
              , ssl
#endif
              , qmail_pop, imap, downloadPercent, deletePercent, debug);

  return popper.doAllWork(connectionsPerMinute);
}

